design-proposals/vm-console-logs.md

```diff
@@ -0,0 +1,743 @@
+```bash
+stirabos@t14s:~$ virtctl console -n simone testvm
+Successfully connected to testvm console. The escape sequence is ^]
+
+login as 'cirros' user. default password: 'gocubsgo'. use 'sudo' for root.
+testvm login: cirros
+Password: 
+$ 
+$ sudo sh -c "echo 'Hello, World!' >> /dev/console"
+Hello, World!
+$ 
+$ exit
+
+login as 'cirros' user. default password: 'gocubsgo'. use 'sudo' for root.
+testvm login: stirabos@t14s:~$ 
+stirabos@t14s:~$ 
+```
+
+and whatever will be written there will also be logged:
+```bash
+stirabos@t14s:~$ kubectl logs -n simone virt-launcher-testvm-t7h26 -c guest-console-log
+[    0.000000] Initializing cgroup subsys cpuset
+[    0.000000] Initializing cgroup subsys cpu
+[    0.000000] Initializing cgroup subsys cpuacct
+[    0.000000] Linux version 4.4.0-28-generic (buildd@lcy01-13) (gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2.1) ) #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 (Ubuntu 4.4.0-28.47-generic 4.4.13)
+...
+login as 'cirros' user. default password: 'gocubsgo'. use 'sudo' for root.
+testvm login: cirros
+Password: 
+$ 
+$ sudo sh -c "echo 'Hello, World!' >> /dev/console"
+Hello, World!
+$ 
+$ exit
+
+login as 'cirros' user. default password: 'gocubsgo'. use 'sudo' for root.
+testvm login: stirabos@t14s:~$ 
+```
+
+## API Examples
+
+On the VM CRD on `/spec/template/spec/domain/devices/autoattachSerialConsole` we already have:
+```yaml
+autoattachSerialConsole:
+  description: Whether to attach the default serial console
+    or not. Serial console access will not be available if
+    set to false. Defaults to true.
+  type: boolean
+```
+to let the VM owner easily attach the default serial console.
+We can introduce an additional boolean `/spec/template/spec/domain/devices/logSerialConsole`:
+```yaml
+logSerialConsole:
+  description: Whether to log the auto-attached default serial
+    console or not. Serial console logs will be collect to
+    a file and then streamed from a named 'guest-console-log'. Not
+    relevant if autoattachSerialConsole is disabled. Defaults
+    to false.
+  type: boolean
+```
+to let the VM owner require the serial console to be logged.
+It should be off by default since serial console logs could potentially disclosure sensitive data.
+
+We should address only the default auto-attached serial console since custom serial device can potentailly contain non-standard settings. If the user really needs something so special to require a special serial console device, we can assume he can also configure the logging there.
+
+On the VM definition side it will be something like:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: testvm
+spec:
+  template:
+    spec:
+      domain:
+        devices:
+          autoattachSerialConsole: true
+          logSerialConsole: true
+...
+```
+
+We can also introduce a cluster wide setting for the cluster-admin to let him tune the default behavior when nothing is specified at VM level.  
+
+## Scalability
```

Comment:
Is there an I/O and CPU cost to this? Especially when enabled on a large cluster (for cluster level configuration as suggested earlier)

Reply:
It really depends on how much each VM will write to its serial console (but this is not so different than for a generic application that produces logs on its stdout).
I'll try to get some numbers for popular guest images.

Reply:
for a small centos-stream8 VM printing 10 log lines every second I see:
```bash
$ kubectl  top pod -n simone virt-launcher-centos-stream8-8v46h --containers --use-protocol-buffers
POD                                  NAME                CPU(cores)   MEMORY(bytes)   
virt-launcher-centos-stream8-8v46h   POD                 0m           0Mi             
virt-launcher-centos-stream8-8v46h   compute             30m          1205Mi          
virt-launcher-centos-stream8-8v46h   guest-console-log   7m           29Mi
```